### ISSUE TITLE
False positive for duplicated code when using special forms

### ISSUE THREAD

**Issue by jcomellas** [rrrene/credo#517]
### Environment

* Credo version: 0.8.10
* Erlang/Elixir version: 1.5.2
* Operating system: Ubuntu 17.10 (AMD64)

### What were you trying to do?
I have three functions in three different modules (that are used to create supervised processes) that are detected as duplicated code. This is happening because, even though the functions correspond to exactly the same text, Credo is not taking into account that the special forms may expand to different values according to the module they are in (e.g. `__MODULE__`).

The function is the following one:
```elixir
  def child_spec(args = [organization_id]) when is_binary(organization_id) do
    %{
      id: __MODULE__,
      start: {__MODULE__, :start_link, [args]},
      restart: :permanent,
      shutdown: 5000,
      type: :worker
    }
  end
```

### Expected outcome
No duplicated code warning or error should be emitted when this situation arises.

### Actual outcome
Credo shows the following errors:
```
  Software Design                                                               
┃ 
┃ [D] ↑ Duplicate code found in lib/audit_log/block_queue.ex:45, 
┃       lib/audit_log/block_writer.ex:52 (mass: 19).
┃       lib/audit_log/event_queue.ex:33 #(AuditLog.EventQueue.child_spec)
┃ [D] ↑ Duplicate code found in lib/audit_log/block_queue.ex:45, 
┃       lib/harmony/event_queue.ex:33 (mass: 19).
┃       lib/harmony/block_writer.ex:52 #(AuditLog.BlockWriter.child_spec)
┃ [D] ↑ Duplicate code found in lib/audit_log/block_writer.ex:52, 
┃       lib/audit_log/event_queue.ex:33 (mass: 19).
┃       lib/audit_log/block_queue.ex:45 #(AuditLog.BlockQueue.child_spec)
```

**Comment by rrrene** [comment]
I think this reveals a deeper problem: Is the "duplicate code" check really benefical for Credo's users?

You could refactor like this, but that would make the code less readable:

```elixir
def child_spec(args = [organization_id]) when is_binary(organization_id) do
  SomeHelperModule.child_spec_for(__MODULE__, args)
end

# SomeHelperModule

def child_spec_for(mod, args) do
  %{
    id: mod,
    start: {mod, :start_link, [args]},
    restart: :permanent,
    shutdown: 5000,
    type: :worker
  }
end
```

This is why the check is "right" in his own mind, but - as I said - I am no longer sure that this is what we want programmers to do ...

WDYT?

**Comment by jcomellas** [comment]
@rrrene IMHO, the duplicated code check for these small functions is not worth it. I know I could refactor my code the way you propose, but these callbacks are used by the supervisor, so removing them would remove important information that should always be visible in the module.

**Comment by rrrene** [comment]
The `DuplicatedCode` check has been moved to the "controversial and experimental checks" of Credo's default config.

Closing this issue. Thx again for this discussion. :+1:
